homeassistant/components/recorder/purge.py

```diff
@@ -44,6 +44,23 @@ def purge_old_data(instance, purge_days: int, repack: bool) -> bool:
 
             _LOGGER.debug("Purging states and events before %s", batch_purge_before)
 
+            # Update old_state_id to NULL before deleting to ensure
+            # the delete does not fail due to a foreign key constraint
+            # since some databases (MSSQL) cannot do the ON DELETE CASCADE
+            # for us.
+            disconnected_rows = (
+                session.query(States)
+                .filter(
```

Comment:
```suggestion
                session.query(States)
                .filter(States.last_updated >= batch_purge_before)
                .filter(
```

It could be enough to only update the remaining (after the purge) states. I think this should work, but to be sure we would need to test it with the different DBMSs. Might be that some do not handle that and don't delete all old states, since at the time of their deletion there was still another old state referencing it.

Reply:
The old state id could be anywhere in the database since it could be days between state changes so I don't think we can limit by `last_updated`.

Reply:
`old_state_id` is always a 1:1 relationship to `state_id`

Reply:
I just tested it with **sqlite** and it works! If both the `old_state` and the `state` referencing it are deleted in the next batch, it's not necessary to set the `old_state_id` of the `state` to none. That's why that change works. This should be a good performance improvement.

Reply:
I'll do some testing with MySQL to make sure it works there as well since it's actually being enforced in this case.

Reply:
@bdraco Sqlite seems to enforce constraints as well. After you mentioned it I did a small test where I excluded the update query (`disconnect_rows`) and nothing was deleted.

More likely that due to some quirk with another DBMS it's not working as it should.
If we where to remove the constraint completely it should definitely be enough.

Reply:
I don't think we can remove it

```
ERROR    homeassistant.components.recorder.util:util.py:43 Error executing query: One or more mappers failed to initialize - can't proceed with initialization of other mappers. Triggering mapper: 'mapped class States->states'. Original exception was: Could not determine join condition between parent/child tables on relationship States.old_state - there are no foreign keys linking these tables.  Ensure that referencing columns are associated with a ForeignKey or ForeignKeyConstraint, or specify a 'primaryjoin' expression.
```

Reply:
I'm going to have to revert that as it breaks MySQL

```
2021-02-19 18:06:46 ERROR (Recorder) [homeassistant.components.recorder.util] Error executing query: (MySQLdb._exceptions.IntegrityError) (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`homeassistant`.`states`, CONSTRAINT `states_ibfk_2` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE NO ACTION)')
[SQL: DELETE FROM states WHERE states.last_updated < %s]
[parameters: (datetime.datetime(2021, 2, 17, 11, 12, 23, tzinfo=<UTC>),)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
2021-02-19 18:06:46 WARNING (Recorder) [homeassistant.components.recorder.purge] Error purging history: (MySQLdb._exceptions.IntegrityError) (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`homeassistant`.`states`, CONSTRAINT `states_ibfk_2` FOREIGN KEY (`old_state_id`) REFERENCES `states` (`state_id`) ON DELETE NO ACTION)')
[SQL: DELETE FROM states WHERE states.last_updated < %s]
[parameters: (datetime.datetime(2021, 2, 17, 11, 12, 23, tzinfo=<UTC>),)]
(Background on this error at: http://sqlalche.me/e/13/gkpj)
```